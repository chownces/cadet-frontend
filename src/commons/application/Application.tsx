import moment from 'moment';
import * as React from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router';

import Academy from '../../pages/academy/AcademyContainer';
import Achievement from '../../pages/achievement/AchievementContainer';
import Contributors from '../../pages/contributors/Contributors';
import Disabled from '../../pages/disabled/Disabled';
import GitHubAssessmentWorkspaceContainer from '../../pages/githubAssessments/GitHubAssessmentWorkspaceContainer';
import GitHubMissionListing from '../../pages/githubAssessments/GitHubMissionListing';
import GitHubCallback from '../../pages/githubCallback/GitHubCallback';
import Login from '../../pages/login/LoginContainer';
import MissionControlContainer from '../../pages/missionControl/MissionControlContainer';
import NotFound from '../../pages/notFound/NotFound';
import Playground from '../../pages/playground/PlaygroundContainer';
import Sicp from '../../pages/sicp/Sicp';
import Sourcecast from '../../pages/sourcecast/SourcecastContainer';
import Welcome from '../../pages/welcome/WelcomeContainer';
import { AssessmentType } from '../assessment/AssessmentTypes';
import NavigationBar from '../navigationBar/NavigationBar';
import Constants from '../utils/Constants';
import { parseQuery } from '../utils/QueryHelper';
import { Role } from './ApplicationTypes';
import { UserCourse } from './types/SessionTypes';

export type ApplicationProps = DispatchProps & StateProps & RouteComponentProps<{}>;

export type DispatchProps = {
  handleLogOut: () => void;
  handleGitHubLogIn: () => void;
  handleGitHubLogOut: () => void;
  updateLatestViewedCourse: (courseId: number) => void;
};

export type StateProps = {
  role?: Role;
  name?: string;
  courses: UserCourse[];
  courseShortname?: string;
  enableAchievements?: boolean;
  enableSourcecast?: boolean;
  assessmentTypes: AssessmentType[];
};

const Application: React.FC<ApplicationProps> = props => {
  const intervalId = React.useRef<number | undefined>(undefined);
  const [isDisabled, setIsDisabled] = React.useState(computeDisabledState());
  const isMobile = /iPhone|iPad|Android/.test(navigator.userAgent);
  const isPWA = window.matchMedia('(display-mode: standalone)').matches; // Checks if user is accessing from the PWA
  const browserDimensions = React.useRef({ height: 0, width: 0 });

  React.useEffect(() => {
    if (Constants.disablePeriods.length > 0) {
      intervalId.current = window.setInterval(() => {
        const disabled = computeDisabledState();
        if (isDisabled !== disabled) {
          setIsDisabled(disabled);
        }
      }, 5000);
    }

    return () => {
      if (intervalId.current) {
        window.clearInterval(intervalId.current);
      }
    };
  }, [isDisabled]);

  /**
   * The following effect prevents the mobile browser interface from hiding on scroll by setting the
   * application height to the window's innerHeight, even after orientation changes. This ensures that
   * the app UI does not break due to the hiding of the browser interface when the user is not on the PWA.
   *
   * Note: When the soft keyboard is up on Android devices, the viewport height decreases and triggers
   * the 'resize' event. The conditional in orientationChangeHandler checks specifically for this, and
   * does not update the application height when the Android keyboard triggers the resize event. IOS
   * devices are not affected.
   */
  React.useEffect(() => {
    const orientationChangeHandler = () => {
      if (
        !(
          window.innerHeight < browserDimensions.current.height &&
          window.innerWidth === browserDimensions.current.width
        )
      ) {
        // If it is not an Android soft keyboard triggering the resize event, update the application height.
        document.documentElement.style.setProperty(
          '--application-height',
          window.innerHeight + 'px'
        );
      }
      browserDimensions.current = { height: window.innerHeight, width: window.innerWidth };
    };

    if (!isPWA && isMobile) {
      orientationChangeHandler();
      window.addEventListener('resize', orientationChangeHandler);
    }

    return () => {
      if (!isPWA && isMobile) {
        window.removeEventListener('resize', orientationChangeHandler);
      }
    };
  }, [isPWA, isMobile]);

  const loginPath = <Route path="/login" render={toLogin(props)} key="login" />;

  const githubAssessmentsPaths = Constants.enableGitHubAssessments
    ? [
        <Route
          path="/githubassessments/missions"
          component={() => (
            <GitHubMissionListing
              handleGitHubLogIn={props.handleGitHubLogIn}
              handleGitHubLogOut={props.handleGitHubLogOut}
            />
          )}
          key="githubAssessmentsMissions"
        />,
        <Route
          path="/githubassessments/editor"
          component={GitHubAssessmentWorkspaceContainer}
          key="githubAssessmentsEditor"
        />
      ]
    : [];

  // Paths for the playground-only deployment
  const playgroundOnlyPaths = [
    <Route path="/playground" component={Playground} key="playground" />,
    <Route path="/contributors" component={Contributors} key="contributors" />,
    <Route path="/callback/github" component={GitHubCallback} key="githubCallback" />,
    <Route exact path="/interactive-sicp" render={redirectToSicp} key="sicpRedirect" />,
    <Route path="/interactive-sicp/:section" component={Sicp} key="sicp" />,
    ...githubAssessmentsPaths
  ];

  // Paths for the Source Academy @NUS deployment
  const fullPaths = [
    loginPath,
    <Route
      path="/playground"
      render={ensureUserAndRouteTo(props, <Playground />)}
      key="authPlayground"
    />,
    ...playgroundOnlyPaths,
    <Route path="/academy" render={toAcademy(props)} key="academy" />,
    <Route path="/welcome" render={ensureUserAndRouteTo(props, <Welcome />)} key="welcome" />
  ];

  if (props.role && props.role !== 'student') {
    fullPaths.push(
      <Route
        path={'/mission-control/:assessmentId(-?\\d+)?/:questionId(\\d+)?'}
        render={toIncubator}
        key="missionControl"
      />
    );
  }

  if (props.enableSourcecast) {
    fullPaths.push(
      <Route
        path="/sourcecast/:sourcecastId?"
        render={ensureUserAndRoleAndRouteTo(props, <Sourcecast />)}
        key="sourcecast"
      />
    );
  }
  if (props.enableAchievements) {
    fullPaths.push(
      <Route
        path="/achievements"
        render={ensureUserAndRoleAndRouteTo(props, <Achievement />)}
        key="achievements"
      />
    );
  }

  const disabled = !['staff', 'admin'].includes(props.role!) && isDisabled;

  const renderDisabled = () => (
    <Disabled reason={typeof isDisabled === 'string' ? isDisabled : undefined} />
  );

  return (
    <div className="Application">
      <NavigationBar
        handleLogOut={props.handleLogOut}
        handleGitHubLogIn={props.handleGitHubLogIn}
        handleGitHubLogOut={props.handleGitHubLogOut}
        updateLatestViewedCourse={props.updateLatestViewedCourse}
        role={props.role}
        name={props.name}
        courses={props.courses}
        courseShortname={props.courseShortname}
        enableAchievements={props.enableAchievements}
        enableSourcecast={props.enableSourcecast}
        assessmentTypes={props.assessmentTypes}
      />
      <div className="Application__main">
        {disabled && (
          <Switch>
            {!Constants.playgroundOnly && loginPath}
            {/* if not logged in, and we're not a playground-only deploy, then redirect to login (for staff) */}
            {!props.role && !Constants.playgroundOnly
              ? [
                  <Route path="/academy" render={redirectToLogin} key={0} />,
                  <Route exact={true} path="/" render={redirectToLogin} key={1} />
                ]
              : []}
            <Route render={renderDisabled} />
          </Switch>
        )}
        {!disabled && Constants.playgroundOnly && (
          <Switch>
            {playgroundOnlyPaths}
            <Route exact={true} path="/" render={redirectToPlayground} />
            <Route component={NotFound} />
          </Switch>
        )}
        {!disabled && !Constants.playgroundOnly && (
          <Switch>
            {fullPaths}
            <Route exact={true} path="/" render={redirectToAcademy} />
            <Route component={NotFound} />
          </Switch>
        )}
      </div>
    </div>
  );
};

const redirectToPlayground = () => <Redirect to="/playground" />;
const redirectToAcademy = () => <Redirect to="/academy" />;
const redirectToLogin = () => <Redirect to="/login" />;
const redirectToWelcome = () => <Redirect to="/welcome" />;
const redirectToSicp = () => <Redirect to="/interactive-sicp/index" />;

/**
 * A user routes to /academy,
 *  1. If the user is logged in, render the Academy component
 *  2. If the user is not logged in, redirect to /login
 */
const toAcademy = ({ name, role }: ApplicationProps) =>
  name === undefined
    ? redirectToLogin
    : role === undefined
    ? redirectToWelcome
    : () => <Academy role={role} />;

/**
 * Routes a user to the specified route,
 *  1. If the user is logged in, render the specified component
 *  2. If the user is not logged in, redirect to /login
 */
const ensureUserAndRouteTo = ({ name }: ApplicationProps, to: JSX.Element) =>
  name === undefined ? redirectToLogin : () => to;

/**
 * Routes a user to the specified route,
 *  1. If the user is logged in and has a latest viewed course, render the
 *     specified component
 *  2. If the user is not logged in, redirect to /login
 *  3. If the user is logged in, but does not have a course, redirect to /welcome
 */
const ensureUserAndRoleAndRouteTo = ({ name, role }: ApplicationProps, to: JSX.Element) =>
  name === undefined ? redirectToLogin : role === undefined ? redirectToWelcome : () => to;

const toLogin = (props: ApplicationProps) => () => {
  const qstr = parseQuery(props.location.search);

  return (
    <Login
      code={qstr.code}
      providerId={qstr.provider}
      providers={[...Constants.authProviders.entries()].map(([id, { name }]) => ({
        id,
        name
      }))}
    />
  );
};

const toIncubator = () => <MissionControlContainer />;

function computeDisabledState() {
  const now = moment();
  for (const { start, end, reason } of Constants.disablePeriods) {
    if (start.isBefore(now) && end.isAfter(now)) {
      return reason || true;
    }
  }
  return false;
}

export default Application;
