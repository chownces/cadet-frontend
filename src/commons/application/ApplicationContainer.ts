import { connect, MapDispatchToProps, MapStateToProps } from 'react-redux';
import { withRouter } from 'react-router';
import { bindActionCreators, Dispatch } from 'redux';

import { logOut } from './actions/CommonsActions';
import { loginGitHub, logoutGitHub, updateLatestViewedCourse } from './actions/SessionActions';
import Application, { DispatchProps, StateProps } from './Application';
import { OverallState } from './ApplicationTypes';

/**
 * Provides the title of the application for display.
 * An object with the relevant properties must be
 * returned instead of an object of type @type {IApplicationProps},
 * as the routing properties of @type {RouteComponentProps} are
 * provided using the withRouter() method below.
 */
const mapStateToProps: MapStateToProps<StateProps, {}, OverallState> = state => ({
  role: state.session.role,
  name: state.session.name,
  courses: state.session.courses,
  courseShortname: state.session.courseShortname,
  enableAchievements: state.session.enableAchievements,
  enableSourcecast: state.session.enableSourcecast,
  assessmentTypes: state.session.assessmentTypes
});

const mapDispatchToProps: MapDispatchToProps<DispatchProps, {}> = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      handleLogOut: logOut,
      handleGitHubLogIn: loginGitHub,
      handleGitHubLogOut: logoutGitHub,
      updateLatestViewedCourse: updateLatestViewedCourse
    },
    dispatch
  );

const ApplicationContainer = withRouter(connect(mapStateToProps, mapDispatchToProps)(Application));

export default ApplicationContainer;
