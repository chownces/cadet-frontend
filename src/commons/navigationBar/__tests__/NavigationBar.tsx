import { shallow } from 'enzyme';

import { Role } from '../../application/ApplicationTypes';
import NavigationBar from '../NavigationBar';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:8000/academy/game'
  })
}));

test('NavigationBar renders "Not logged in" correctly', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    courses: [],
    assessmentTypes: []
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});

test('NavigationBar renders correctly for student with course', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    courses: [
      {
        courseId: 1,
        courseName: 'CS1101S Programming Methodology (AY20/21 Sem 1)',
        courseShortName: 'CS1101S',
        viewable: true
      }
    ],
    courseId: 1,
    courseShortName: 'CS1101S',
    enableAchievements: true,
    enableSourcecast: true,
    role: Role.Student,
    name: 'Bob',
    assessmentTypes: ['Missions', 'Quests', 'Paths', 'Contests', 'Others']
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});

test('NavigationBar renders correctly for student without course', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    courses: [],
    name: 'Bob',
    assessmentTypes: []
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});
