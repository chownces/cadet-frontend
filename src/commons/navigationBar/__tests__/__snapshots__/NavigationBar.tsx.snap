// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar renders "Not logged in" correctly 1`] = `
"<Fragment>
  <Blueprint3.Navbar className=\\"NavigationBar primary-navbar bp3-dark\\">
    <Blueprint3.NavbarGroup align=\\"left\\">
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/academy\\">
        <Blueprint3.Icon icon=\\"symbol-diamond\\" />
        <Blueprint3.NavbarHeading style={{...}}>
          Source Academy
        </Blueprint3.NavbarHeading>
      </NavLink>
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/sourcecast\\">
        <Blueprint3.Icon icon=\\"music\\" />
        <div className=\\"navbar-button-text\\">
          Sourcecast
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/playground\\">
        <Blueprint3.Icon icon=\\"code\\" />
        <div className=\\"navbar-button-text\\">
          Playground
        </div>
      </NavLink>
    </Blueprint3.NavbarGroup>
    <Blueprint3.NavbarGroup align=\\"right\\">
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/contributors\\">
        <Blueprint3.Icon icon=\\"heart\\" />
        <div className=\\"navbar-button-text hidden-sm hidden-xs\\">
          Contributors
        </div>
      </NavLink>
      <div className=\\"visible-xs\\">
        <Blueprint3.NavbarDivider className=\\"thin-divider\\" />
      </div>
      <div className=\\"hidden-xs\\">
        <Blueprint3.NavbarDivider className=\\"default-divider\\" />
      </div>
      <Dropdown handleLogOut={[Function: handleLogOut]} name={[undefined]} />
    </Blueprint3.NavbarGroup>
  </Blueprint3.Navbar>
</Fragment>"
`;

exports[`NavigationBar renders correctly with student role 1`] = `
"<Fragment>
  <Blueprint3.Navbar className=\\"NavigationBar primary-navbar bp3-dark\\">
    <Blueprint3.NavbarGroup align=\\"left\\">
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/academy\\">
        <Blueprint3.Icon icon=\\"symbol-diamond\\" />
        <Blueprint3.NavbarHeading style={{...}}>
          Source Academy
        </Blueprint3.NavbarHeading>
      </NavLink>
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/sourcecast\\">
        <Blueprint3.Icon icon=\\"music\\" />
        <div className=\\"navbar-button-text\\">
          Sourcecast
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/playground\\">
        <Blueprint3.Icon icon=\\"code\\" />
        <div className=\\"navbar-button-text\\">
          Playground
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/achievement\\">
        <Blueprint3.Icon icon=\\"mountain\\" />
        <div className=\\"navbar-button-text\\">
          Achievement
        </div>
      </NavLink>
    </Blueprint3.NavbarGroup>
    <Blueprint3.NavbarGroup align=\\"right\\">
      <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link bp3-button bp3-minimal\\" to=\\"/contributors\\">
        <Blueprint3.Icon icon=\\"heart\\" />
        <div className=\\"navbar-button-text hidden-sm hidden-xs\\">
          Contributors
        </div>
      </NavLink>
      <Blueprint3.NavbarDivider className=\\"default-divider\\" />
      <Blueprint3.Tooltip content=\\"Toggle Telebay\\" position=\\"bottom\\" hoverCloseDelay={0} hoverOpenDelay={100} minimal={false} transitionDuration={100}>
        <Blueprint3.Button onClick={[Function: onClick]} icon=\\"compass\\" minimal={true} style={{...}} />
      </Blueprint3.Tooltip>
      <div className=\\"visible-xs\\">
        <Blueprint3.NavbarDivider className=\\"thin-divider\\" />
      </div>
      <div className=\\"hidden-xs\\">
        <Blueprint3.NavbarDivider className=\\"default-divider\\" />
      </div>
      <Dropdown handleLogOut={[Function: handleLogOut]} name={[undefined]} />
    </Blueprint3.NavbarGroup>
  </Blueprint3.Navbar>
  <AcademyNavigationBar role=\\"student\\" />
</Fragment>"
`;

exports[`NavigationBarMobileSideMenu renders "Not logged in" correctly 1`] = `
"<Blueprint3.Drawer isOpen={true} position=\\"left\\" onClose={[Function: onClose]} title=\\"Telebay\\" icon=\\"compass\\" className=\\"bp3-dark\\" canOutsideClickClose={true} style={{...}} vertical={false}>
  <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" to=\\"/sourcecast\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"music\\" />
    <div>
      Sourcecast
    </div>
  </NavLink>
  <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" to=\\"/playground\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"code\\" />
    <div>
      Playground
    </div>
  </NavLink>
</Blueprint3.Drawer>"
`;

exports[`NavigationBarMobileSideMenu renders correctly with student role 1`] = `
"<Blueprint3.Drawer isOpen={true} position=\\"left\\" onClose={[Function: onClose]} title=\\"Telebay\\" icon=\\"compass\\" className=\\"bp3-dark\\" canOutsideClickClose={true} style={{...}} vertical={false}>
  <NavLink to=\\"/academy/missions\\" activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"flame\\" />
    <div>
      Missions
    </div>
    <Connect(NotificationBadge) notificationFilter={[Function (anonymous)]} disableHover={true} />
  </NavLink>
  <NavLink to=\\"/academy/quests\\" activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"lightbulb\\" />
    <div>
      Quests
    </div>
    <Connect(NotificationBadge) notificationFilter={[Function (anonymous)]} disableHover={true} />
  </NavLink>
  <NavLink to=\\"/academy/paths\\" activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"predictive-analysis\\" />
    <div>
      Paths
    </div>
    <Connect(NotificationBadge) notificationFilter={[Function (anonymous)]} disableHover={true} />
  </NavLink>
  <NavLink to=\\"/academy/contests\\" activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"comparison\\" />
    <div>
      Contests
    </div>
    <Connect(NotificationBadge) notificationFilter={[Function (anonymous)]} disableHover={true} />
  </NavLink>
  <NavLink to=\\"/academy/practicals\\" activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"manual\\" />
    <div>
      Others
    </div>
  </NavLink>
  <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" to=\\"/sourcecast\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"music\\" />
    <div>
      Sourcecast
    </div>
  </NavLink>
  <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" to=\\"/playground\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"code\\" />
    <div>
      Playground
    </div>
  </NavLink>
  <NavLink activeClassName=\\"bp3-active\\" className=\\"NavigationBar__link__mobile bp3-button bp3-minimal bp3-large\\" to=\\"/achievement\\" onClick={[Function: onClose]}>
    <Blueprint3.Icon icon=\\"mountain\\" />
    <div>
      Achievement
    </div>
  </NavLink>
</Blueprint3.Drawer>"
`;
